### PR TITLE
[swift] change `unknownFields` type to `[UInt32: Data]` from just `Data`.

### PR DESCRIPTION
**Pull request by dnkoutso**
This change improves the already difficult usability of unknown fields but also prepares for a follow up change to re-do Wire proto extensions for Swift.

This is a breaking change that will ship with 5.x of Wire which is a major version bump anyway.

### CODE REVIEW

File: wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift [hunk 1]
```diff
@@ -229,14 +229,14 @@ public final class ProtoReader {
             )
         }
         
-        //nil out unknownFields to avoid memory leaks caused by
+        //clear out unknownFields to avoid memory leaks caused by
         //UnsafeMutablePointer's inability to deallocate non-trivial types.
         //This is caused by unknownFields type, WriteBuffer, being a class/reference type.
         defer {
-            messageStack[token].unknownFields = nil
+            messageStack[token].unknownFields.removeAll()
```
Comment (dnkoutso) [inline wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift:236]: @lickel does the comment still apply as is?

Comment (lickel) [reply wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift:236]: Probably?

File: wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift [hunk 2]
```diff
@@ -209,14 +209,14 @@ public final class ProtoReader {
         return nil
     }
 
-    public func endMessage(token: Int) throws -> Data {
+    public func endMessage(token: Int) throws -> [UInt32: Data] {
```
Comment (lickel) [inline wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift:212]: You should create that type alias now and use that here

File: wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift [hunk 3]
```diff
@@ -551,8 +551,10 @@ public final class ProtoWriter {
     // MARK: - Public Methods - Unknown Fields
 
     /** Append unknown fields data to the output */
-    public func writeUnknownFields(_ data: Data) throws {
-        self.buffer.append(data)
+    public func writeUnknownFields(_ fields: [UInt32: Data]) throws {
```
Comment (lickel) [inline wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift:554]: Typealias

File: wire-runtime-swift/src/main/swift/wellknowntypes/Duration.swift [hunk 4]
```diff
@@ -79,7 +79,7 @@ public struct Duration {
      * to +999,999,999 inclusive.
      */
     public var nanos: Int32
-    public var unknownFields: Foundation.Data = .init()
+    public var unknownFields: [UInt32 : Foundation.Data] = .init()
```
Comment (lickel) [inline wire-runtime-swift/src/main/swift/wellknowntypes/Duration.swift:82]: Typealias

File: wire-runtime-swift/src/test/swift/ProtoReaderTests.swift [hunk 5]
```diff
@@ -1263,7 +1281,7 @@ extension ProtoReader {
         return value!
     }
 
-    func forEachTag(_ decode: (UInt32) throws -> Void) throws -> Foundation.Data {
+    func forEachTag(_ decode: (UInt32) throws -> Void) throws -> [UInt32: Foundation.Data] {
```
Comment (lickel) [inline wire-runtime-swift/src/test/swift/ProtoReaderTests.swift:1284]: Typealias

**Review by lickel — APPROVED** [review]
- [ ] Probably should get a second approval here

**Review by swankjesse — APPROVED** [review]
Looks good as far as Wire the library is concerned.

**Comment by swankjesse** [comment]
Have ya seen how we track unknown fields in Java/Kotlin? We keep a ByteString with the encoded data for everything we couldn’t decode.

**Comment by dnkoutso** [comment]
I just did a quick check, certainly seems nicer to what we have in Swift right now. Is it more about the immutability, utility and performance to adopt something equivalent in Swift?

**Pull request merged**